### PR TITLE
chore: add contract test testcases generator

### PR DESCRIPTION
**Pull request by OmarAlJarrah**

### CODE REVIEW

File: .github/workflows/generate-contract-tests.yaml [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+      configurations: -th -te ${{ github.event.inputs.endpoint_prefix }} --operationIdsToTags --oneOf
+  generate:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/download-artifact@v4
+        with:
+          name: specs
+          path: sdk-test
+      - name: Set up JDK 17
```
Comment (mohnoor94) [inline .github/workflows/generate-contract-tests.yaml]: 21

File: .github/workflows/generate-contract-tests.yaml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        uses: actions/setup-java@v4
+        with:
+          java-version: '17'
+          distribution: 'temurin'
+      - name: Generate Contract Tests
+        working-directory: sdk-test
+        run:
+          mkdir -p contract-tests
+          output=$(realpath ./contract-tests)
+          mvn clean install exec:java -Dexec.args="-i specs.yaml -o $output"
```
Comment (mohnoor94) [inline .github/workflows/generate-contract-tests.yaml]: What is  "output" here?

File: .github/workflows/generate-contract-tests.yaml [hunk 3]
```diff
@@ -0,0 +1,55 @@
+name: Generate Contract Tests
+on:
+  workflow_call:
+    inputs:
+      specs_url:
```
Comment (mohnoor94) [inline .github/workflows/generate-contract-tests.yaml]: Please use the new self-serve actions.

File: sdk-test/src/main/kotlin/com/expediagroup/sdk/test/CLI.kt [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+        contractTestsGenerator =
+            ContractTestsGenerator(
+                spec = spec,
+                outputDir = File(outputDir),
+                maxTestCombinations = maxTestCombinations
+            )
+
+        contractTestsGenerator.generate(writeMode = generateSdkTests).also { testCases ->
+            if (generateSdkTests) {
```
Comment (mohnoor94) [inline sdk-test/src/main/kotlin/com/expediagroup/sdk/test/CLI.kt]: Let's use the early exit if here (`if not --> return`)

File: sdk-test/src/main/kotlin/com/expediagroup/sdk/test/contract/Constant.kt [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ * limitations under the License.
+ */
+package com.expediagroup.sdk.test.contract
+
+val SPECMATIC_CONFIG =
+    mapOf(
+        "host" to "localhost",
+        "port" to "8080",
+        "SPECMATIC_GENERATIVE_TESTS" to "true",
+        "MAX_TEST_REQUEST_COMBINATIONS" to "1"
```
Comment (mohnoor94) [inline sdk-test/src/main/kotlin/com/expediagroup/sdk/test/contract/Constant.kt]: Looks like this is not needed.

File: sdk-test/src/main/kotlin/com/expediagroup/sdk/test/contract/function/ExtractPathParamsFunc.kt [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        request: HttpRequest
+    ): Map<String, Any> =
+        mutableMapOf<String, Any>().apply {
+            val originalPath = scenario.path.split("/")
+            val resolvedPath = request.path?.split("/") ?: emptyList()
+
+            if (originalPath.size.equals(resolvedPath.size).not()) {
+                return@apply
+            }
+//            omar/google/(company:string)
```
Comment (mohnoor94) [inline sdk-test/src/main/kotlin/com/expediagroup/sdk/test/contract/function/ExtractPathParamsFunc.kt]: ???

File: sdk-test/src/main/kotlin/com/expediagroup/sdk/test/contract/model/api/TestCaseHttpResponse.kt [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+import com.fasterxml.jackson.annotation.JsonProperty
+import io.specmatic.core.HttpResponse
+
+data class TestCaseHttpResponse(
+    @JsonProperty val status: Int
+) {
+    companion object {
+        fun from(response: HttpResponse): TestCaseHttpResponse =
+            TestCaseHttpResponse(
+                status = response.status
```
Comment (mohnoor94) [inline sdk-test/src/main/kotlin/com/expediagroup/sdk/test/contract/model/api/TestCaseHttpResponse.kt]: Shouldn't we compare the actual responses? (The data).

File: sdk-test/src/main/kotlin/com/expediagroup/sdk/test/openapi/SdkTestGenerator.kt [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    private val namespace: String,
+    private val spec: File,
+    private val version: String = "1.0.0",
+    private val testCases: List<TestCaseApiCall> = emptyList()
+) {
+    val product = Product(namespace, "kotlin")
+    val config =
+        CodegenConfigurator().apply {
+            setGeneratorName("kotlin")
+            setTemplateDir("templates/expediagroup-sdk")
```
Comment (mohnoor94) [inline sdk-test/src/main/kotlin/com/expediagroup/sdk/test/openapi/SdkTestGenerator.kt]: Let's take this as input.

File: sdk-test/src/main/kotlin/com/expediagroup/sdk/test/openapi/SdkTestGenerator.kt [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+            }
+
+            addAdditionalProperty(CodegenConstants.API_PACKAGE, product.apiPackage)
+            addAdditionalProperty(CodegenConstants.ENUM_PROPERTY_NAMING, "UPPERCASE")
+            addAdditionalProperty(CodegenConstants.LIBRARY, "jvm-ktor")
+            addAdditionalProperty(CodegenConstants.SERIALIZATION_LIBRARY, "jackson")
+            addAdditionalProperty(CodegenConstants.SORT_PARAMS_BY_REQUIRED_FLAG, true)
+
+            // Template specific properties
+            addAdditionalProperty("shadePrefix", product.shadePrefix)
```
Comment (mohnoor94) [inline sdk-test/src/main/kotlin/com/expediagroup/sdk/test/openapi/SdkTestGenerator.kt]: Let's remove the unneeded properties.

**Review by mohnoor94 — CHANGES_REQUESTED** [review]

File: sdk-test/src/main/kotlin/com/expediagroup/sdk/test/CLI.kt (hunk 4 above)
Comment (OmarAlJarrah) [reply sdk-test/src/main/kotlin/com/expediagroup/sdk/test/CLI.kt]: Done.

File: sdk-test/src/main/kotlin/com/expediagroup/sdk/test/contract/Constant.kt (hunk 5 above)
Comment (OmarAlJarrah) [reply sdk-test/src/main/kotlin/com/expediagroup/sdk/test/contract/Constant.kt]: Done.

File: .github/workflows/generate-contract-tests.yaml (hunk 1 above)
Comment (OmarAlJarrah) [reply .github/workflows/generate-contract-tests.yaml]: Removed the action as it no longer is needed.

File: .github/workflows/generate-contract-tests.yaml (hunk 2 above)
Comment (OmarAlJarrah) [reply .github/workflows/generate-contract-tests.yaml]: Removed the action as it no longer is needed.

File: .github/workflows/generate-contract-tests.yaml (hunk 3 above)
Comment (OmarAlJarrah) [reply .github/workflows/generate-contract-tests.yaml]: Removed the action as it no longer is needed.

File: sdk-test/src/main/kotlin/com/expediagroup/sdk/test/contract/function/ExtractPathParamsFunc.kt (hunk 6 above)
Comment (OmarAlJarrah) [reply sdk-test/src/main/kotlin/com/expediagroup/sdk/test/contract/function/ExtractPathParamsFunc.kt]: Removed.

File: sdk-test/src/main/kotlin/com/expediagroup/sdk/test/contract/model/api/TestCaseHttpResponse.kt (hunk 7 above)
Comment (OmarAlJarrah) [reply sdk-test/src/main/kotlin/com/expediagroup/sdk/test/contract/model/api/TestCaseHttpResponse.kt]: Done.

File: sdk-test/src/main/kotlin/com/expediagroup/sdk/test/openapi/SdkTestGenerator.kt (hunk 8 above)
Comment (OmarAlJarrah) [reply sdk-test/src/main/kotlin/com/expediagroup/sdk/test/openapi/SdkTestGenerator.kt]: Done,

File: sdk-test/src/main/kotlin/com/expediagroup/sdk/test/openapi/SdkTestGenerator.kt (hunk 9 above)
Comment (OmarAlJarrah) [reply sdk-test/src/main/kotlin/com/expediagroup/sdk/test/openapi/SdkTestGenerator.kt]: Done.

File: sdk-test/src/main/kotlin/com/expediagroup/sdk/test/contract/model/api/TestCaseHttpRequest.kt [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+
+    /**
+     * Checks if the HTTP request has a non-empty, non-blank body.
+     *
+     * @return `true` if the body is not empty and not just a JSON object, `false` otherwise.
+     */
+    fun hasBody(): Boolean =
+        toBoolean(body, dynamic = true).and(!(body!!.isEmptyJsonObject()))
+
+    /**
```
Comment (anssari1) [inline sdk-test/src/main/kotlin/com/expediagroup/sdk/test/contract/model/api/TestCaseHttpRequest.kt]: Remove dangling comments

File: sdk-test/src/main/kotlin/com/expediagroup/sdk/test/contract/model/api/TestCaseHttpRequest.kt [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+        }
+    }
+
+    /**
+     * Retrieves the body of the HTTP request if it is not empty and not just a JSON object.
+     *
+     * @return The body of the request if it is not empty and not a default empty JSON object,
+     *         otherwise `null`.
+     */
+    fun getRequestBody(): String? = if (hasBody()) body!! else null
```
Comment (anssari1) [inline sdk-test/src/main/kotlin/com/expediagroup/sdk/test/contract/model/api/TestCaseHttpRequest.kt]: Remove redundant !!

File: sdk-test/src/main/resources/templates/expediagroup-sdk/executor.mustache [hunk 12]
```diff
@@ -0,0 +1,59 @@
+package com.expediagroup.sdk.{{{namespace}}}.test;
+
+
```
Comment (anssari1) [inline sdk-test/src/main/resources/templates/expediagroup-sdk/executor.mustache]: Replace wild card imports

File: sdk-test/src/main/kotlin/com/expediagroup/sdk/test/contract/model/api/TestCaseHttpRequest.kt (hunk 11 above)
Comment (OmarAlJarrah) [reply sdk-test/src/main/kotlin/com/expediagroup/sdk/test/contract/model/api/TestCaseHttpRequest.kt]: ✅

File: sdk-test/src/main/kotlin/com/expediagroup/sdk/test/contract/model/api/TestCaseHttpRequest.kt (hunk 10 above)
Comment (OmarAlJarrah) [reply sdk-test/src/main/kotlin/com/expediagroup/sdk/test/contract/model/api/TestCaseHttpRequest.kt]: ✅

File: sdk-test/src/main/resources/templates/expediagroup-sdk/executor.mustache (hunk 12 above)
Comment (OmarAlJarrah) [reply sdk-test/src/main/resources/templates/expediagroup-sdk/executor.mustache]: ✅

**Review by anssari1 — APPROVED** [review]

**Review by mohnoor94 — APPROVED** [review]
reviewed by Nasser

**Pull request merged**